### ISSUE TITLE
Can we force warming of compiler by compiling dummy file

### ISSUE THREAD

**Issue by eed3si9n** [sbt/sbt#2984]
Rory reported that when running on batch mode `sbt compile`, the first run on the machine is faster than the second onwards. His speculation was that compilation of compile bridge could have forced JITing on the first run.
Would it make sense to run a small dummy file (something like Hello world) before the `compile` on startup?

**Comment by jvican** [comment]
JVM will compile concrete paths of the compiler. Compiling a dummy Scala file is not sufficient to get the compiler hot. In fact, it may  evenintroduce a performance regression since the JVM will optimize concrete paths, and while compiling a completely different project those paths will change and, supposing that the C2 has done some work, most of it will be thrown away because JIT will deoptimize paths. This would only work if you compile the same project at the beginning, and that wouldn't make a lot of sense.
  
I lean towards letting the user know that Scala compiler should always be kept hot. I don't think sbt can do anything to make compilation hot.

**Comment by dwijnand** [comment]
@jvican One really hard-core min-max idea that we had was to derive a good minimal source file, representative of the code paths in the compiler the real project used. To more accurately warm up the compiler.

I would put this one in the "experiment here, because there might be good wins in some cases" box.

> I lean towards letting the user know that Scala compiler should always be kept hot.

That's being tracked as #2987. This derives from the same story from Rory, but it's distinct.

**Comment by jvican** [comment]
> One really hard-core min-max idea that we had was to derive a good minimal source file, that was representative of the code paths in the compiler the real compiler used. To more accurately warm up the JVM.
  
In my opinion, it's impossible to get such a source file for any Scala codebase. I would really not like that sbt does this warming up the compiler for me. Also, it takes way more than one compiler iteration to properly warm up Scalac.
  
I think my issue is that a build tool should not take care of this. I find it a very bad idea, it's unpredictable and when I type `compile`, I really mean "just compile my project", nothing else.
  
No other JVM-based build tool does so. I suggest reconsidering the scope of this ticket, and even closing it. In any case, this is not something that can be done in a hackathon of 4 hours, since it requires deep knowledge of both scalac and JVM.

**Comment by dwijnand** [comment]
My suggestion was an extreme extension to the original issue, which I mentioned because you were going into details about how a simple hello world wouldn't be sufficient. But let's discuss your feedback:

> Also, it takes way more than one compiler iteration to properly warm up Scalac.

There's no restriction on iterations. If more than one provides better overall compile speeds then do more than one.

> I think my issue is that a build tool should not take care of this. I find it a very bad idea, it's unpredictable and when I type compile, I really mean "just compile my project", nothing else.

You're looking for scalac then.

> No other JVM-based build tool does so. I suggest reconsidering the scope of this ticket, and even closing it.

 We ran 1 example and it _did_ improve overall compile time. So it's worth pursuing.

> In any case, this is not something that can be done in a hackathon of 4 hours, since it requires deep knowledge of both scalac and JVM.

Agreed. It's niche and specific.

**Comment by jvican** [comment]
I think this is a bad idea for the reasons explained above. I'm repeating them:
* JIT will compile concrete paths. Those paths will get decompiled when other projects that differ from the seed Scala file get compiled by the same class loader, which is going to be very likely. Deoptimization takes time and this can slow down compile times.
* To get a really hot Scalac, you need more than one or two iterations. If sbt is going to do this, which I think it's a bad idea, I see why it doesn't compile that seed file X times until compiler is hot.
* Build tools are no JVMs. They should let the runtime they run on optimize their code. Conceptually, sbt should not do anything more than just `compile`.
  
That said, I'm no expert in the JVM. I know that @DarkDimius has an extensive experience on this, so perhaps he can share his mind on this topic.

**Comment by DarkDimius** [comment]
> JVM will compile concrete paths of the compiler. Compiling a dummy Scala file is not sufficient to get the compiler hot.

Depends what you mean here. You're oversimplifying by assuming there are two states: cold(interpreted) and hot(compiled). There's a lot more possible states which you could either start from or end up with.

A slightly less simplified model would be that there are several compilation levels: interpreted, C1(client compiler) and C2(server compiler). Both of compilers have some kind of budget for compilation. This budget is a combination of: 
 - method size;
 - compilation time;
 - number of times this method has been recompiled.

The VM starts by running code interpreted, then jumps it to C1 and than may compile it with C2. Compilation with C2 is very costly but provides very good code, while C1 is fast and provides decent code. More details(and a less simplified model) can be found in my comment here https://github.com/lampepfl/dotty/issues/222

Now, considering the idea of warmup runs. Warmup is a good thing when you want vm to be stable. Stable doesn't always mean `fast`. Due to deoptimizations, you could have a situation where first runs where faster than the next ones. Warmup can be a bad thing if we make VM do optimization that it will have to revert, as this can lead to
 - using methods compiled during warmup that could have been re-compiled more efficiently;
 - jit deciding that there has been to many deoptimizations and deciding not to apply some optimizations.
It is true that due to warmup that doesn't match the actual behavior you could make it worse.

Now, we're talking about warming up the compiler here. Compiler is a strange beast that already behaves very unfriendly to VM, as has code that runs in several modes, that can loosely be called "typer", "pre-erasure" and "after-erasure". What I've seen in dotty because of this is that during the compilation of big projects JIT tends to over-optimize for "typer" first, than we get hit by a lot of de-optimization when transformations start. Same happens near erasure. This doesn't happen for "hello world", as it's not compiled long enough to trigger costly JIT optimizations in many places.

What we should also consider is that there may be more things in play here. E.g. if you pre-fetch data that you may need during real compilation, you'll save time otherwise spend loading it. You've already have them in your disk cache or in your memory. This helps a lot.

To illustrate dotty, REPL pre-warms itself by compiling and evaluating
`val theAnswerToLifeInTheUniverseAndEverything = 21 * 2`. // you can actually access this variable from repl :-)
Why we did it? 
 - to ensure that commonly used paths in compiler are compiled. Compiling this expression is not enough to trigger to much of costly optimizations in JIT and it means we don't over-optimize for it;
 - to ensure that we've pre-loaded common parts of Scala and Java std libraries that compiled needs during compilation.


I think doing something along those lines could be a good idea for SBT. I would expect that compiling something bigger for warmup will not help much, but may be harmful instead.

**Comment by DarkDimius** [comment]
Note that it only make sense to do if sbt is either idling or is willing to trade startup time.

**Comment by jvican** [comment]
> You're oversimplifying by assuming there are two states: cold(interpreted) and hot(compiled). There's a lot more possible states which you could either start from or end up with.
  
Yeah, I'm definitely oversimplifying. The discussion about this ticket was that Scalac would be hot and that would be beneficial, so my points followed the same line of thought and level of abstraction. I didn't know what you do in Dotty, that's cool!
  
> I think doing something along those lines could be a good idea for SBT. I would expect that compiling something bigger for warmup want help much, but may be harmful instead. Note that it only make sense to do if sbt is either idling or is willing to trade startup time.
  
Exactly, but this proves my point. Compiling a hello world file right before doing `compile` is bad, because the time X for compiling Hello world and the user-defined project will always be equal or higher than Y where Y is the raw compilation time of the user-defined project. This would only make sense if you do this before the user runs `compile` without sacrificing reload time.
  
What @DarkDimius  seems to be proposing is to compile hello world without the user knowing it, on the background, where the cpu is idle, which is definitely not what the description of this ticket says. This could backfire if there are other processes that, because of their intensive use of memory, evict all the cache entries added by the Hello World extra compilation. I would like to add that, because of NUMA, I think that this would only work if the core compiling hello world is the same as the core compiling the user-defined project.
  
It would be great that if this is done, we have JMH benchmarks results to confirm that this solution buy us something.

**Comment by dwijnand** [comment]
> I think doing something along those lines could be a good idea for SBT. I would expect that compiling something bigger for warmup will not help much, but may be harmful instead.

Thank you kindly @DarkDimius for such an in-depth explanation. For comparison this was the "hello world" I was playing with:

```scala
package t

object Hello {
  def main(args: Array[String]): Unit = ()
}
```

(lol, just noticed it doesn't even print hello world)

@jvican 

> Compiling a hello world file right before doing compile is bad, because the time X for compiling Hello world and the user-defined project will always be equal or higher than Y where Y is the raw compilation time of the user-defined project.

My shoddy little test would appear to indicate that that isn't true.

> It would be great that if this is done, we have JMH benchmarks results to confirm that this solution buy us something.

You know my position on this. It's absolutely a requirement for such a change, to me.

**Comment by eed3si9n** [comment]
It takes 3s to compile Hello world. If the compiler is 1.3x ~ 2x faster, then most build would benefit from this preemptive warming. Dale already demonstrated that last night with Akka Actor - https://twitter.com/dwijnand/status/837613953719107584
